structcol/montecarlo.py

```diff
@@ -184,57 +233,69 @@ def scatter(self, sintheta, costheta, sinphi, cosphi):
 
             kz = -sintheta[n-1,:] * cosphi[n-1,:] * denom + kn[2,n-1,:] * costheta[n-1,:]
 
-            # if kz is close to 1 and >= 0:
+            # If kz is close to 1 (propagation direction is straight down in 
+            # z-axis) and >= 0, the set of equatons are [1]:
             kx2 = sintheta[n-1,:] * cosphi[n-1,:]
             ky2 = sintheta[n-1,:] * sinphi[n-1,:]
             kz2 = costheta[n-1,:]
 
-            # if kz is close to 1 and < 0:
+            # If kz is close to 1 and < 0, the set of equatons are [1]:
             ky3 = -sintheta[n-1,:] * sinphi[n-1,:]
             kz3 = -costheta[n-1,:]
 
-            kn[:,n,:] = np.where(denom > eps, (kx, ky, kz),
+            # Choose which set of equations to use based on the conditions
+            # described above
+        
+            kn[:,n,:] = np.where(np.logical_not(np.isclose(1-denom**2, 1, rtol=1e-05, atol=1e-08, equal_nan=False)), (kx, ky, kz),
                                  np.where(kn[2,n-1,:]>= 0., (kx2, ky2, kz2), (kx2, ky3, kz3)))
+            
+            # (KEEP UNTIL FURTHER REVISIONS)
```

Comment:
always delete stale code.  It is stored in previous revisions in github, and it's best to avoid accumulation of cruft.